File: app/src/main/java/com/timewasteanalyzer/usage/control/UsageRepository.java
package com.timewasteanalyzer.usage.control;


import android.app.usage.UsageEvents;
import android.app.usage.UsageStatsManager;
import android.content.Context;

import com.timewasteanalyzer.usage.model.AppUsage;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;


public class UsageRepository {

    private List<AppUsage> mAppUsageList;
    private UsageStatsManager mUsageStatsManager;
    private Context mContext;

    private HashMap<String, AppUsage> appUsageMap;
    private List<UsageEvents.Event> allEvents;

    private long phoneUsageTotal;

    private static UsageRepository sInstance;

    /**
     * Private constructor only called in Singleton initliazer {@link #getInstance(Context)}.
     *
     * @param context the current {@link Context}
     */
    private UsageRepository(Context context) {
        mContext = context;
        mUsageStatsManager = (UsageStatsManager) context.getSystemService(Context.USAGE_STATS_SERVICE);
    }

    /**
     * Singleton initializer.
     *
     * @param context the current {@link Context}.
     * @return the singleton instance of this class
     */
    public static UsageRepository getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new UsageRepository(context);

        }
        return sInstance;
    }

    /**
     * Queries the usage data for the passed {@link FilterType}.
     *
     * @param filterType the current {@link FilterType}
     */
    public void queryUsageStatisticsForType(FilterType filterType) {
        resetFormerData();

        long now = System.currentTimeMillis();
        long startTime = 0;
        switch (filterType) {
            case DAY:
                // TODO filter from 0:00 instead of 24h
                startTime = now - 1000 * 3600 * 24; // querying last day
                break;

            case WEEK:
                // TODO test whether last weeks events are queried like this
                startTime = now - 1000 * 3600 * 24 * 7;
                break;

            case ALL:
                // TODO test whether all events are queried like this
                startTime = 0;
                break;
        }

        // Query events and initialize repository data
        assert mUsageStatsManager != null;
        UsageEvents queriedUsage = mUsageStatsManager.queryEvents(startTime, now);
        initializeDataBasedOn(queriedUsage);

        // Iterate through usage list
        for (int i = 0; i < allEvents.size() - 1; i++) {
            UsageEvents.Event event = allEvents.get(i);
            UsageEvents.Event event2 = allEvents.get(i + 1);

            // Calculate opened counter
            if (!event.getPackageName()
                      .equals(event2.getPackageName()) && event2.getEventType() == 1) {
                // if true, E1 (launch event of an app) app launched
                appUsageMap.get(event2.getPackageName())
                           .increaseLaunchCount();
            }

            // UsageTime of apps in time range
            if (event.getEventType() == 1 && event2.getEventType() == 2 && event.getClassName()
                                                                                .equals(event2.getClassName())) {
                // Update total phone usage
                long diff = event2.getTimeStamp() - event.getTimeStamp();
                phoneUsageTotal += diff;

                // Update current usage
                AppUsage currentUsage = appUsageMap.get(event.getPackageName());
                currentUsage.addTimeInForeground(diff);
                currentUsage.updatePercentage(phoneUsageTotal);
            }
        }
        updateUsageList(appUsageMap.values());
    }

    private void resetFormerData() {
        phoneUsageTotal = 0;
        appUsageMap = new HashMap();
        allEvents = new ArrayList<>();
    }

    /**
     * Resets the {@link #appUsageMap} and {@link #allEvents}. Afterwards the data is updated based on the passed
     * {@link UsageEvents}.
     *
     * @param usageEvents the queried {@link UsageEvents}
     */
    private void initializeDataBasedOn(UsageEvents usageEvents) {
        UsageEvents.Event currentEvent;
        while (usageEvents.hasNextEvent()) {
            currentEvent = new UsageEvents.Event();
            usageEvents.getNextEvent(currentEvent);
            if (currentEvent.getEventType() == UsageEvents.Event.MOVE_TO_FOREGROUND || currentEvent.getEventType() ==
                    UsageEvents.Event.MOVE_TO_BACKGROUND) {
                allEvents.add(currentEvent);

                // Save usages into map
                String key = currentEvent.getPackageName();
                if (appUsageMap.get(key) == null)
                    appUsageMap.put(key, new AppUsage(mContext, key));
            }
        }
    }

    private void updateUsageList(Collection<AppUsage> values) {
        mAppUsageList.clear();
        mAppUsageList.addAll(values);
        mAppUsageList.sort((usage1, usage2) -> {
            long time1 = usage1.getMsInForeground();
            long time2 = usage2.getMsInForeground();
            return time1 < time2 ? 1 : -1;
        });
    }

    public String getTotalTypeForFilter() {
        return getTodayHeading();
    }

    public List<AppUsage> getUsageList() {
        if (mAppUsageList == null) {
            mAppUsageList = new ArrayList<>();
        }
        return mAppUsageList;
    }

    public String getTodayHeading() {
        int seconds = (int) (phoneUsageTotal / 1000) % 60;
        int minutes = (int) ((phoneUsageTotal / (1000 * 60)) % 60);
        int hours = (int) ((phoneUsageTotal / (1000 * 60 * 60)) % 24);

        return new StringBuilder().append("Today: ")
                                  .append(String.format("%02d:%02d:%02d", hours, minutes, seconds))
                                  .toString();
    }
}

File: app/src/main/kotlin/com/timewasteanalyzer/usage/control/UsageRepository.kt
package com.timewasteanalyzer.usage.control


import android.app.usage.UsageEvents
import android.app.usage.UsageStatsManager
import android.content.Context
import com.timewasteanalyzer.R
import com.timewasteanalyzer.usage.model.AppUsage
import com.timewasteanalyzer.util.SingletonHolder
import java.time.LocalDateTime
import java.util.*

class UsageRepository private constructor(context: Context) {

    private var mAppUsageList: MutableList<AppUsage> = ArrayList()
    private var mUsageStatsManager: UsageStatsManager
    private var mAppUsageMap: HashMap<String, AppUsage> = HashMap()
    private var mAllEvents: MutableList<UsageEvents.Event> = ArrayList()
    private var mContext: Context = context
    private var phoneUsageTotal: Long = 0
    private var mCurrentFilterType: FilterType

    init {
        mContext = context
        mUsageStatsManager = mContext.getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
        mCurrentFilterType = FilterType.DAY
    }

    companion object : SingletonHolder<UsageRepository, Context>(::UsageRepository)

    val getTotalTimeHeading: String
        get() {
            var filterText = when (mCurrentFilterType) {
                FilterType.DAY -> R.string.title_today
                FilterType.WEEK -> R.string.title_week
            }

            val seconds = (phoneUsageTotal / 1000).toInt() % 60
            val minutes = (phoneUsageTotal / (1000 * 60) % 60).toInt()
            val hours = (phoneUsageTotal / (1000 * 60 * 60) % 24).toInt()

            return StringBuilder()
                    .append(mContext.getString(filterText))
                    .append(": ")
                    .append(String.format("%02d:%02d:%02d", hours, minutes, seconds))
                    .toString()
        }

    val mUsageList: List<AppUsage>
        get() {
            return mAppUsageList
        }

    /**
     * Queries the usage data for the passed [FilterType].
     *
     * @param filterType the current [FilterType]
     */
    fun queryUsageStatisticsForCurrentType() {
        resetFormerData()

        val now = System.currentTimeMillis()
        var startTime: Long = when (mCurrentFilterType) {
            // querying hours of current day
            FilterType.DAY -> now - 1000 * 3600 * LocalDateTime.now().hour

            FilterType.WEEK -> now - 1000 * 3600 * 24 * 7
        }

        // Query events and initialize repository data
        val queriedUsage = mUsageStatsManager.queryEvents(startTime, now)
        mUsageStatsManager.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, 0, now)
        initializeDataBasedOn(queriedUsage)

        // Iterate through usage list
        for (i in 0 until mAllEvents.size - 1) {
            val event = mAllEvents[i]
            val event2 = mAllEvents[i + 1]

            // Calculate opened counter
            if (event.packageName != event2.packageName && event2.eventType == 1) {
                // if true, E1 (launch event of an app) app launched
                mAppUsageMap[event2.packageName]?.increaseLaunchCount()
            }

            // UsageTime of apps in time range
            if (event.eventType == 1 && event2.eventType == 2 && event.className == event2.className) {
                // Update total phone usage
                val diff = event2.timeStamp - event.timeStamp
                phoneUsageTotal += diff

                // Update current usage
                val currentUsage = mAppUsageMap[event.packageName]
                currentUsage?.addTimeInForeground(diff)
            }
        }

        // Percentages have to updated after total phone usage is calculate based on all events.
        mAppUsageMap.values.forEach { it.updatePercentage(phoneUsageTotal) }

        updateUsageList(mAppUsageMap.values)
    }

    private fun resetFormerData() {
        phoneUsageTotal = 0
        mAppUsageMap = HashMap()
        mAllEvents = ArrayList()
    }

    /**
     * Resets the [.mAppUsageMap] and [.mAllEvents]. Afterwards the data is updated based on the passed
     * [UsageEvents].
     *
     * @param usageEvents the queried [UsageEvents]
     */
    private fun initializeDataBasedOn(usageEvents: UsageEvents) {
        var currentEvent: UsageEvents.Event
        while (usageEvents.hasNextEvent()) {
            currentEvent = UsageEvents.Event()
            usageEvents.getNextEvent(currentEvent)
            if (currentEvent.eventType == UsageEvents.Event.MOVE_TO_FOREGROUND || currentEvent.eventType == UsageEvents.Event.MOVE_TO_BACKGROUND) {
                mAllEvents.add(currentEvent)

                // Save usages into map
                val key = currentEvent.packageName
                if (mAppUsageMap[key] == null)
                    mAppUsageMap[key] = AppUsage(mContext, key)
            }
        }
    }

    private fun updateUsageList(values: Collection<AppUsage>) {
        mAppUsageList.clear()
        mAppUsageList.addAll(values.sortedWith(compareBy { -it.msInForeground }))
    }

    /**
     * Setter for {@link #mCurrentFilterType}.
     */
    fun setCurrentType(currentFilter: FilterType) {
        mCurrentFilterType = currentFilter
    }

    /**
     * Deletes all entries from {@link #mAppUsageList}.
     */
    fun clear() {
        mAppUsageList.clear()
    }
}